### ISSUE TITLE
BM25 search issues

### ISSUE THREAD

**Issue by zoltan-fedor** [semi-technologies/weaviate#2455]
I was working on the Haystack codebase to restore its compatibility with Weaviate when it comes to BM25 searches and I hit an interesting issue - making me believe that something might be wrong with the BM25 implementation in Weaviate, so started an investigation to the BM25 search in Weaviate. Below are the findings.

As I was adding Weaviate to the Haystack BM25 unit tests, the same tests using some example records to search with the BM25 search that have worked with other databases like Elasticsearch and a direct in-memory implantation of BM25, has kept failing to return the right result for Weaviate.
It was clear that this is not a Haystack issue, so I have just made [a special search query for Weaviate in the Haystack tests and added a comment about Weaviate being funny here](https://github.com/deepset-ai/haystack/pull/3723/files#diff-a41bf841d7722c38e95f8b93caab4d938ee526ab858a27fe9fd71d5898178d82R63).

But this still means that there is an issue with the BM25 search seemingly working a lot less effectively than the one in other databases, so I spent time to reproduce this issue directly in Weaviate (without Haystack), so I can bring this to the attention of the Weaviate team.

Below is what I found.

I have created a class in Weaviate with a single `text` property and a single record with the value `My name is Carla and I live in Berlin` in that text field and ran the following BM25 searches to see which one manages to return that record:

- query: `Who lives in Berlin` => no result  (this query DOES return a result in all other BM25 databases, which is what triggered this investigation)
- query: `Carla, I lives in Berlin` => no result
- query: `Carla, I live in Berlin` => record has been found
- query: `I live in Berlin` => record has been found
- query: `Who live in Berlin` => record has been found
- query: `live` => record has been found
- query: `lives` => no result
- query: `Carla Berlin` => no result
- query: `carla Berlin` => record has been found

These results seem to point to multiple issues.

When searching the `My name is Carla and I live in Berlin`  record with `Carla, I lives in Berlin`, there is no result, while multiple words should match, like `Carla`, `Berlin` and `lives` (this last one if the words in the query do get stemmed (as they should be in BM25)).

Testing the stemming directly: search query `live` brings back the result, but `lives` does not, proving that no stemming is performed.

Also no result for `Carla Berlin`, but there are results for `carla Berlin`, proving that the original text ( `My name is Carla and I live in Berlin`) is lowercased when indexed, but the query running on that index is not (hence `carla` and `Carla` is no match).
That is obviously wrong, clearly a bug, as now we are comparing apples to oranges. Sure, we could ask the users to lowercase all their queries for BM25, but if we don't ask them to lowercase the data when inserted into the class, then we shouldn't ask them to do that when querying the same record (and also it is not required in other BM25-compatible databases to lowercase the query).

I hope these issues can be looked at before with would consider the BM25 functionality production ready, which I believe is on the roadmap for `v1.17`

---------------------------------

The code to reproduce my findings:
```
import weaviate

client = weaviate.Client(url='http://127.0.0.1:8080', timeout_config = (10, 60)) 

# create a test class
class_schema = {
    "class": "BM25Tester",
    "description": "Just a simple class to test BM25 queries against",
    "properties": [
        {
            "name": "content",
            "dataType": ["text"],
            "description": "The content we will be searching",
        }
    ]
}
client.schema.create_class(class_schema)

# add the single record
client.data_object.create(
    class_name='BM25Tester',
    data_object = {'content': 'My name is Carla and I live in Berlin'}
)

# run the search you want to test
r = client.query.raw("""{
  Get {
    BM25Tester (
        bm25: {
          query: "Who lives in Berlin",
          properties: ["content"]
        }
    ) {
      content
    }
  }
}"""
)
print(r)
```

**Comment by dandv** [comment]
I can confirm that stemming was not intended to be functional at this point in the preview, but case insensitivity should work. We'll take a deeper look.

CC @donomii

**Comment by zoltan-fedor** [comment]
Yes, I was wondering whether the lack of stemming was intentional, so I was just reading up on whether stemming is a mandatory or a nice-to-have part of BM25.

It seems that stemming is definitely not a mandatory part of BM25, but also found it to be almost always highly productive to include stemming with BM25 - [see this popular paper on BM25](http://www.cs.otago.ac.nz/homepages/andrew/papers/2014-2.pdf), which is likely why most BM25 implementations include it.

While the case sensitivity is being looked at, as stemming was not intended to be functional at this point in the preview, may I ask when it is expected to be included? Should I be raising a dedicate issue for adding stemming to BM25, so it can be added to the roadmap and users can vote for it? Thanks!

**Comment by etiennedi** [comment]
Issue for stemming: https://github.com/semi-technologies/weaviate/issues/2439

Case sensitivity (pre-processing) in general, should be fixed in the latest builds, e.g: https://github.com/semi-technologies/weaviate/actions/runs/3706835320#summary-10123516261

**Comment by zoltan-fedor** [comment]
Hah, it seems that you beat me to raising the stemming issue by 2 days! :-)

Let me give it a try to that latest build, so I can do a quick test on the case sensitivity to confirm. If it gets confirmed, then this ticket can get closed, as the stemming already has its ticket.

**Comment by zoltan-fedor** [comment]
Thanks @etiennedi, I have tested the latest build and I can confirm that build 1.17.0-prealpha-e23d5fd did solve the case sensitivity issue I have raised in this ticket.

The other issue I have raised here is related to the lack of stemming, for which you have already raised a ticket #2439 , so this issue can be closed now.

Thanks for the help and looking forward to seeing the stemming being implemented from the roadmap!
